### PR TITLE
add 'ipfs repo migrate' command

### LINKED ISSUE #7471: ipfs migration without starting the daemon
Currently, it seems to be impossible to do a migration (if necessary) without starting the daemon. 

I'm a package maintainer and like script the migration after installation of the new package instead of running the migration on the first run of the daemon, which - as discussed earlier in a PR - might hit the timeout of systemd for starting a service.

I like to have an `ipfs migrate --assume-yes` or something like this for unattended migration while updating a package with a package manager.

It should return 0 if everything is fine, regardless if a migration was necessary or not.

### PR DESCRIPTION
**Pull request by zaibon**
this command allows to run the repo migration without starting the daemon.

resolves #7471

### CODE REVIEW

**Comment by aschmahmann** [comment]
Thanks @zaibon for starting this. Could you do two more things:

1) As specified in the issue you linked we need an `--allow-downgrade` flag. While this command has one it's not hooked up to anything.
2) There is nothing testing that this command works. You can modify https://github.com/ipfs/go-ipfs/blob/83c57eda1fe4c50660761b401b4e4f016e6ecfd0/test/sharness/t0066-migration.sh to also test that `ipfs repo migrate` and `ipfs repo migrate --allow-downgrade` work as expected

Thanks again and let me know if you need any guidance here.

**Comment by zaibon** [comment]
Thanks for the review @aschmahmann.

> As specified in the issue you linked we need an --allow-downgrade flag. While this command has one it's not hooked up to anything.

Right, forgot that part. What should be the behavior when this flag is set ? At the moment I always call `migrate.RunMigration(fsrepo.RepoVersion)` should I read the desired version from somewhere else and then compare the current version with the desired version and only allow downgrade if the flag is set ?

> There is nothing testing that this command works. You can modify https://github.com/ipfs/go-ipfs/blob/83c57eda1fe4c50660761b401b4e4f016e6ecfd0/test/sharness/t0066-migration.sh to also test that ipfs repo migrate and ipfs repo migrate --allow-downgrade work as expected

Yep, will do, thanks for the guidance here.

File: test/sharness/t0066-migration.sh
Comment (zaibon) [inline test/sharness/t0066-migration.sh:16]: Had to change this here to make the detection of the `fs-repo-migrations` to work since it seems now the code also checks the version the migration tool is able to handle.

**Comment by zaibon** [comment]
@aschmahmann I hooked up the `allow-downgrade` flag. For this I had to add an extra argument to the `RunMigration` method. Not sure this is ok, I'm usually not a big fan of boolean argument. What's thought about it ?

I also added some test for the new command. I'm not sure how I can test the downgrade though, some input there would be welcome :-)

**Comment by aschmahmann** [comment]
@zaibon sorry it's taking a while to get to this. A little busy with things related to the v0.7.0 release, but will try and get to this next week.

**Comment by zaibon** [comment]
@aschmahmann ping ;-)

**Comment by aschmahmann** [comment]
Thanks @zaibon for the reminder and apologies it slipped off my radar. I'll try and take a look over the weekend.

I was actually thinking about this PR yesterday because some of the pain around migrations has made implementing something like https://github.com/ipfs/fs-repo-migrations/issues/98 more important. This would likely rely on `ipfs repo migrate` instead of a separate `fs-repo-migrations` binary which would make this command even more useful (and make testing it simpler too).

**Comment by zaibon** [comment]
Nice, I'll have a look at https://github.com/ipfs/fs-repo-migrations/issues/98.

Should I already try to rebase this PR since I guess it's a bit out of date with master or wait your review ?

**Comment by aschmahmann** [comment]
> Should I already try to rebase this PR since I guess it's a bit out of date with master or wait your review 

I suspect it's an easy rebase, but if it's a pain for some reason then I'd hold off until there's some progress on ipfs/fs-repo-migrations#98. If we do basically what's recommended in that issue then this command will replace the fs-repo-migrations binary for people doing manual migration manipulation.

**Comment by gammazero** [comment]
@zaibon Wanted to let you know that this PR has not been forgotten about.  The work to [overhaul the migrations and binary distribution](https://github.com/ipfs/fs-repo-migrations/issues/98#issuecomment-763329183) is code-complete, and is waiting until there is time to review it all. Once #7857 is merged, then you can rebase this PR and make any necessary changes.  Hopefully in the next couple of weeks.

**Comment by zaibon** [comment]
@gammazero seems https://github.com/ipfs/fs-repo-migrations/issues/98 is now closed. Can I start rebasing and make the required change here or is there anything else in the pipeline ?

**Comment by gammazero** [comment]
@zaibon You can rebase now.  You will need to make some changes to how you call `RunMigration`.  See here:
https://github.com/ipfs/go-ipfs/blob/master/cmd/ipfs/daemon.go#L291-L293

**Comment by aschmahmann** [comment]
@zaibon not sure if this is ready for re-review yet, but when you're ready just request review from me or tag me

**Comment by gammazero** [comment]
@zaibon I took a look at the changes to the sharness tests, and realized that what was there (before your changes) needs to be fixed.  go-ipfs no longer uses `fs-repo-migrations` to perform migrations, and the tests were not doing what they were supposed to be doing.

Attn: @aschmahmann 
I have created a PR #8053 to fix this.  It includes comments explaining what is going on as well.  Please rebase this PR as soon as it can be merged. (sorry for another roadblock).

**Comment by zaibon** [comment]
@aschmahmann I just rebased on top of #8053, so this PR should be ready now.

File: test/sharness/t0066-migration.sh
Comment (gammazero) [inline test/sharness/t0066-migration.sh:93]: This looks like a duplicate of the test at line 72-74.

You may want to take this test from the current master branch and reapply your changes.

File: test/sharness/t0066-migration.sh
Comment (gammazero) [inline test/sharness/t0066-migration.sh:97]: The variable `IPFS_REPO_VER` holds the original version, so you should be able to do:
```sh
echo "$IPFS_REPO_VER" > "$IPFS_PATH"/version
```
This way when there is a new repo version, this test will not need to be updated.

**Review by gammazero — CHANGES_REQUESTED** [review]

**Comment by gammazero** [comment]
@aschmahmann With the recent changes to allow migrations over IPFS, and to import downloaded migrations, we probably want this PR to at least be able to fetch migrations via IPFS.  This command would also benefit from being able to import downloaded migrations as CAR files.

**Comment by aschmahmann** [comment]
> we probably want this PR to at least be able to fetch migrations via IPFS. 

👍

> This command would also benefit from being able to import downloaded migrations as CAR files.

We get this basically for free if we can fetch the migrations using IPFS then if the user really wants to work with CAR files they can use `ipfs dag import <migration car>` before they run `ipfs repo migrate` and that'll do the job.

File: test/sharness/t0066-migration.sh
Comment (zaibon) [reply test/sharness/t0066-migration.sh:93]: This test is up to date with master branch, all the new code is only related this me PR.

This line looks similar to 71-74, cause it does test the same thing but it is the result of another command. In 71-74 we test the result of the `ipfs daemon` command, while here we test the result of the `ipfs migrate` command.

**Comment by zaibon** [comment]
@aschmahmann Do you still need some action from me on this ?

**Comment by gammazero** [comment]
@zaibon IPFS repo migrations has been a moving target, and there is additional functionality that your command should include.  This is the ability to download migrations from IPFS and from gateways according to a new Migrations section of the IPFS config.  What this means for this PR is the following:
1. Wait for PR #8153 to merge
2. Rebase this PR on master
3. Replace your code [here](https://github.com/ipfs/go-ipfs/pull/7658/files#diff-daf81b7f18381e7ad219f36448f52d0db26222ec838be8dfc37629694dc682e5R403-R404) with code based on the example [here](https://github.com/ipfs/go-ipfs/blob/chore/expose-migration-api/cmd/ipfs/daemon.go#L296-L315)

**Comment by BigLep** [comment]
@zaibon : can you please rebase so we can merge this?

**Comment by zaibon** [comment]
@BigLep done.

**Comment by Matchjuan859** [comment]
am not sure wwhATA U NED ME TO DUE

**Comment by gammazero** [comment]
I took care of [step 3 above](https://github.com/ipfs/go-ipfs/pull/7658#issuecomment-848194387) to allow this command to use IPFS to download migrations.

**Comment by BigLep** [comment]
Arg, and not it looks like CI is not passing.  I have limited internet bandwidth to track it down, but I believe I sawt his mentioned somewhere else.  @aschumann is there another issue we can track against so can then merge this?

**Comment by BigLep** [comment]
@gammazero : can you please rebase on master and see if it's good to go?

**Comment by gammazero** [comment]
@aschmahmann @BigLep As far as I can tell, this PR is done.  The problem is that it does not have privileges to run the sharness tests using a `2xlarge` instance as configured in the `.circleci/config.yml` [here](https://github.com/ipfs/go-ipfs/blob/master/.circleci/config.yml#L124), and this is causing failure.

**Comment by gammazero** [comment]
Closing: This PR is replaced by #8428 to move it into a branch to avoid CI problems associated with the repo.

**Pull request closed without merging**
